### PR TITLE
MAINT: parametrize scipy/sparse/linalg/_isolve/tests/test_iterative.py

### PR DESCRIPTION
**Pull request by h-vetinari**
Pulled out test parametrization from https://github.com/scipy/scipy/pull/18391, based on #18462, planning to rebase once that's in.

I turned them into proper parametrized fixtures to avoid having to duplicate the list of solvers around all over the place. Should be reviewed per commit, and ideally without showing whitespace-only changes.

CC @ilayn

### CODE REVIEW

File: scipy/sparse/linalg/_isolve/tests/test_iterative.py [hunk 1]
```diff
@@ -507,55 +509,49 @@ def test_x0_working(solver):
...
-        count = [0]
-        x, info = solver(A, b, callback=cb, show=True)
-        out, err = capsys.readouterr()
-        if i == 20:  # Asymmetric and Positive Definite
-            assert_equal(out, f"{solverstring}: Linear solve not converged "
-                              f"due to reach MAXIT iterations {count[0]}\n")
-        else:  # 1-D Poisson equations
-            assert_equal(out, f"{solverstring}: Linear solve converged due to "
-                              f"reach TOL iterations {count[0]}\n")
-        assert_equal(err, '')
```
Comment (h-vetinari) [inline scipy/sparse/linalg/_isolve/tests/test_iterative.py:555]: If I read this test correctly, it iterates through the cases and has a very specific error expectation for exactly the 20th Case. I counted and hope that that's actually `"rand-cmplx-sym-pd-F"`. I'm also not sure about the necessity of the callback here at all (and how the counter through the cases has anything to do with the iterations), but for now I thought I'd keep passing one.

Comment (h-vetinari) [reply scipy/sparse/linalg/_isolve/tests/test_iterative.py]: Aside from counting incorrectly (off by one), the analysis was correct. :)

**Comment by h-vetinari** [comment]
> Looks much better but still won't fix the actual problem there.

I'm not claiming to solve any problem here, it's explicitly meant to be NFC (non-functional change). It's meant to reduce the necessity for test changes in https://github.com/scipy/scipy/pull/18391

**Comment by h-vetinari** [comment]
In the spirit of "if you care strongly, I'll concede", I'm happy to back out https://github.com/scipy/scipy/pull/18463/commits/31fd62303c0e6a9b5a10c08da21c33c4158775c6 if anyone objects. It's based on a single [change](https://github.com/scipy/scipy/pull/18462#discussion_r1193329134) that I had picked up from #18391 and applied to the whole file.

**Comment by ilayn** [comment]
This is much better than what I did in the PR anyways because I basically got tired by looking at failing tests for long time. So +1 to merge from me.

**Comment by h-vetinari** [comment]
Glad to hear it @ilayn.

@tupui, do you want to review (not required, just an offer). I'm still still planning to remove `self.Poisson1D` & `2D` from the class, which I overlooked previously, but that's only two lines. I'll do it once I'm back at a computer in about an hour

**Comment by h-vetinari** [comment]
@tupui, not sure if your 👍 means you want to review or you're happy as is.

**Comment by tupui** [comment]
It means I m happy as is 😃

**Review by j-bowhay — COMMENTED** [review]
LGTM too, thanks @h-vetinari

**Pull request merged**